lib/puppet/provider/puppet_certificate/ruby.rb

```diff
@@ -1,127 +1,191 @@
 require 'fileutils'
 require 'puppet/face'
+require 'puppet/ssl/certificate'
 
 Puppet::Type.type(:puppet_certificate).provide(:ruby) do
-  desc "Manage Puppet certificates using the certificate face"
+  desc 'Manage Puppet certificates'
+
+  def initialize(value = {})
+    super(value)
+    if at_least_puppet6
+      @cert_provider = Puppet::X509::CertProvider.new(hostprivkey: nil, hostcert: nil)
+      @machine = Puppet::SSL::StateMachine.new
+    else
+      debug('using legacy code and APIs')
+    end
+  end
 
   def create
     debug "create #{@resource[:name]}"
-    generate_key
     submit_csr
-    if ca_location == 'local'
-      sign_certificate
-    else
-      retrieve_certificate
-    end
+    sign_certificate if @resource[:ca_location] == :local
+    retrieve_certificate
   end
 
-  def generate_key
-    unless key
-      debug "generating new key for #{@resource[:name]}"
-      ensure_cadir if ca_location == 'local'
-      Puppet::SSL::Oids.register_puppet_oids
+  def submit_csr
+    debug "generating new key for #{@resource[:name]} and submitting csr"
+    ensure_cadir if @resource[:ca_location] == :local
+    Puppet::SSL::Oids.register_puppet_oids if Puppet::SSL::Oids.respond_to?('register_puppet_oids')
```

Comment:
are there situations where the method isn't available?

Reply:
Yes.  In old versions the method didn't exist, but that's ok as core puppet called it for you.